Feature: 소셜 로그인 구현(카카오, 네이버, 구글)

Choi-Jiwon-38 opened the pull request:
## #️⃣연관된 이슈

#1 

## 📝작업 내용

> 클라이언트
> - custom axios instance의 baseURL 오탈자 수정
> - 카카오 JS SDK 추가
> - 카카오 간편 로그인 연결(인가 코드 발급)
> - 네이버 간편 로그인 연결

> 서버
> - `/auth/kakao` -  카카오 access token 발급 및 사용자 정보 조회
> - `/auth/naver` - 네이버 access token 발급 및 사용자 정보 조회
> - 각 소셜 로그인 이후, 회원 유무에 따라 회원가입, 로그인 분기 처리

### 스크린샷


## 💬리뷰 요구사항

> - 로컬 환경에서 클라이언트, 서버를 둘 다 실행시켜야지 확인이 가능합니다.
> - `localhost:3000/signin`에서 변경 사항을 테스트할 수 있습니다.

Choi-Jiwon-38 committed:
feat: 카카오 간편 로그인 추가(client, js sdk 이용)

- 클라이언트 측에서 카카오 로그인을 위한 인가 코드 전달 부분까지 구현 완료
- 카카오 간편 로그인 기능을 위해서 REST API가 아닌 JS SDK 사용

Choi-Jiwon-38 committed:
fix: `baseURL` 오탈자 수정

Choi-Jiwon-38 committed:
feat: 카카오 토큰 발급 API 추가

Choi-Jiwon-38 committed:
feat: 카카오 토큰 발급 API 연결

- 클라이언트에서 서버 측으로 인가 코드를 통하여 카카오 토큰을 발급하는 API 연결
- 인가 코드를 통한 토큰 발급은 1회성이기 때문에 useRef를 이용하여 여러 번 요청이 가지 않도록 처리

daun-up commented on frontend/src/app/signin/page.tsx:
```diff
@@ -0,0 +1,52 @@
+declare global {
+  interface Window {
+    Kakao: any;
+    naver: any;
```
근데 any 써도 되나요!!!!!!

Choi-Jiwon-38 replied:
이 부분은 제가 나중에 타입 잡으려고 하다가 놓친 부분 같습니다. 확인해주셔서 감사합니다! 해당 부분은 수정해두겠습니다 :)

frontend-jiwon-choi replied:
해당 부분은 따로 라이브러리를 추가하거나 라이브러리를 참고해서 kakao-js-sdk에서 사용 단계에서 필요한 type만 잡을 수 있을 정도로 직접 타입을 추가해두어도 괜찮을 거 같습니다. 이건 시간나면 제가 진행할게요.

https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/kakao-js-sdk

Choi-Jiwon-38 replied:
해당 부분은 따로 라이브러리를 추가하거나 라이브러리를 참고해서 kakao-js-sdk에서 사용 단계에서 필요한 type만 잡을 수 있을 정도로 직접 타입을 추가해두어도 괜찮을 거 같습니다. 이건 시간나면 제가 진행할게요.

https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/kakao-js-sdk

VarGun commented on frontend/src/app/signin/kakao/page.tsx:
```diff
@@ -0,0 +1,35 @@
+  const code = searchParams.get('code');
+  /**
+   * @description 카카오 액세스 토큰 요청 시, 발생하는 'KOE320 invalid_grant' 에레 방지용
+   * / 동일한 인가 코드를 두 번 이상 사용하지 못하도록 useRef를 flag처럼 사용합니다.
```
추후에 strict 모드 조정해서 isGranted 는 없애는 방향으로 진행하는 것도 좋아보입니다.

Choi-Jiwon-38 replied:
현재는 StrictMode 때문에 2번 요청가서 막아둔 건 맞는데, 추후에도 해당 `useEffect`를 통한 액세스 토큰 요청이 2번 갈 수도 있는 부분 때문에 남겨두긴 했습니당. 전체 코드 완성 이후에도 `useEffect` 호출이 한 번만 이뤄지는 게 확인되면 제거하도록 할게요!

Choi-Jiwon-38 committed:
Merge branch 'develop' of https://github.com/KMU-WINK/workinkorea into feature/1-social-accounts

Choi-Jiwon-38 commented on frontend/src/app/signin/page.tsx:
```diff
@@ -0,0 +1,52 @@
+    }
+  };
+
+  const handleKakaoLoad = () => {
+    if (window.Kakao) {
+      window.Kakao.init(process.env.NEXT_PUBLIC_KAKAO_JAVASCRIPT_KEY);
+    }
+  };
+
+  useEffect(() => {
+    // 회원가입 및 로그인 페이지에 진입할 때 Kakao SDK를 추가
+    const kakaoScript = document.createElement('script');
+    kakaoScript.src = 'https://developers.kakao.com/sdk/js/kakao.min.js';
+    kakaoScript.async = true;
+    kakaoScript.onload = handleKakaoLoad;
+    document.body.appendChild(kakaoScript);
+
+    return () => {
+      // 회원가입 및 로그인 페이지에서 이탈할 때 Kakao SDK를 제거
+      document.body.removeChild(kakaoScript);
+    };
+  }, []);
```
@daun-up 

[Next.js Script Optimization 문서](https://nextjs.org/docs/app/building-your-application/optimizing/scripts)을 참고해서 해당 부분을 `useEffect`를 통하여 mount 시에 스크립트를 추가하고 unmount 시에 스크립트를 제거하는 것이 아니라 layout에 Script를 사용하는 것으로 변경 부탁 드리겠습니다 :)

이 부분이 수정이 완료되면 위 `loginWithKakao`에서도 if (window.Kakao)를 이용한 별도의 체크를 안해도 될 것으로 보여요.

Choi-Jiwon-38 committed:
Merge branch 'develop' of https://github.com/KMU-WINK/workinkorea into feature/1-social-accounts

Choi-Jiwon-38 committed:
refactor: `Next.js` 권장 사항에 맞추어 `Script` 분리

- 기존 `useEffect`를 사용하여 mount, unmount 시에 추가 및 제거하였던 script를 RootLayout의 Script로 리팩토링

daun-up committed:
feat: naver, google 로그인 추가

Choi-Jiwon-38 commented on frontend/src/services/auth/naver.ts:
```diff
@@ -0,0 +1,36 @@
+  code,
+  state,
+}: TAuthorizeWithNaver) => {
+  const response = await PublicAxiosInstance.post(
+    'https://nid.naver.com/oauth2.0/token',
+    null, // POST 요청에서 body가 없는 경우
+    {
+      params: {
+        grant_type: 'authorization_code',
+        client_id: clientId,
+        client_secret: clientSecret,
+        redirect_uri: redirectUri,
+        code,
+        state,
+      },
+    },
+  );
```
@daun-up 

이 부분 token 발급이 post가 아니라 get이에요!
그런데 네이버의 access token 발급은 클라이언트에서 전부 수행할 수가 없는 것 같습니다.

(slack DM으로 공유주셨던 문제도 200 OK가 떨어졌지만 에러가 발생한 이유는 이 부분 같아요)
```
naver:1 Access to XMLHttpRequest at 'https://nid.naver.com/oauth2.0/token?grant_type=authorization_code&client_id=vM1KI0Xu2BN5DZ9DUGIA&client_secret=NFTmyuzgMw&redirect_uri=http:%2F%2Flocalhost:3000%2Fsignin%2Fnaver&code=tGJeoCgiK5xvfqgvwO&state=random_string_for_security' from origin 'http://localhost:3000' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

그래서 아래와 같은 두 가지 방법 중 한 가지를 선택하여 수정할 수 있을 것 같습니다.

> 1. 네이버 로그인 authorize 요청에서의 redirect_url을 백엔드 서버쪽으로 넘겨서 토큰 발급을 백엔드에서 진행
> 2. 현재 프론트 /signin/naver의 search param으로 받은 코드를 백엔드 서버쪽으로 넘겨서 토큰 발급을 백엔드에서 진행

결론적으로는 네이버 API에서 CORS 관련 헤더 설정을 안해둔 상황이어서 저희가 토큰 발급을 백엔드에서 작업해야 해요. 첫 번째 방법이 보안상 좀 더 안전하다고 합니다 :)

Choi-Jiwon-38 commented on frontend/src/app/signin/page.tsx:
```diff
@@ -16,13 +19,13 @@ const SignInPage = () => {
-        <button>네이버 로그인</button>
-      </div>
-      <button>구글 로그인</button>
+      <button type="button" onClick={loginWithKaKao}>
+        카카오 로그인
+      </button>
+      <NaverLoginButton />
+      <GoogleLoginButton />
```
@daun-up 

button 태그에 `type="button"` 추가해주신 것 좋습니다!

Choi-Jiwon-38 commented on frontend/src/app/signin/naver/page.tsx:
@daun-up 

다른 코멘트에 네이버 API CORS 관련하여 작성해두었는데 짧게 요약하자면 토큰 발급을 서버 쪽에서 처리해야 에러를 피할 수 있을 것 같아요. 그래서 방식을 바꾸면 해당 콜백 페이지는 없애도 될 것 같습니다.

Choi-Jiwon-38 commented on frontend/src/app/signin/_components/NaverLogin/index.tsx:
```diff
@@ -0,0 +1,22 @@
+function NaverLoginButton() {
+  const handleLogin = () => {
+    const clientId = process.env.NEXT_PUBLIC_NAVER_CLIENT_ID;
+    const redirectUri = `${window.location.protocol}//${window.location.host}/signin/naver`;
```
@daun-up 

이 부분이 백엔드 서버쪽 URI로 바뀔 거 같아요! 참고 정도만 해주세요 :)

Choi-Jiwon-38 committed:
fix: argument 전달 오류 수정

Choi-Jiwon-38 committed:
fix: 네이버 access token 발급을 서버에서 요청하도록 수정

Choi-Jiwon-38 committed:
Merge branch 'develop' of https://github.com/KMU-WINK/workinkorea into feature/1-social-accounts

Choi-Jiwon-38 committed:
test: 소셜 토큰 발급 후 사용자 조회 및 분기 처리 추가(테스트 필요)

Choi-Jiwon-38 committed:
refactor: 인가코드, 토큰 발급 서버에서 처리 / 프론트 callback 페이지 제거

Choi-Jiwon-38 committed:
Merge branch 'develop' of https://github.com/KMU-WINK/workinkorea into feature/1-social-accounts

Choi-Jiwon-38 committed:
fix: 별도 패키지 없이 jwt token 발급하도록 수정

- JWT_SECRET_KEY는 추가 예정
- 현재 User 관련 테이블 에러 발생하여 이 부분 확인 필요

Choi-Jiwon-38 committed:
hotfix: `id`의 type이 `String`이어서 User 테이블 변경 필요

Choi-Jiwon-38 committed:
remove: google 로그인 콜백 페이지 미사용 예정

Choi-Jiwon-38 committed:
Merge branch 'develop' of https://github.com/KMU-WINK/workinkorea into feature/1-social-accounts

Choi-Jiwon-38 committed:
remove: merge 과정에서 중복 코드 제거

Choi-Jiwon-38 committed:
Merge branch 'develop' of https://github.com/KMU-WINK/workinkorea into feature/1-social-accounts

Choi-Jiwon-38 committed:
refactor: `social_id` field 추가 반영

Choi-Jiwon-38 commented on backend/app/routers/auth.py:
```diff
@@ -0,0 +1,94 @@
+@router.get("/kakao")
+async def kakaoAuth(code: str):
+    client_id = os.getenv('KAKAO_REST_API_KEY') 
+    redirect_uri = 'http://127.0.0.1:8000/auth/kakao'
```
이 부분 서버 쪽 url이 들어가야 하는데 제가 하드코딩 해두었네요.. 이 부분은 수정하겠습니다.

Choi-Jiwon-38 committed:
feat: 소셜 로그인 작업 사항 병합(로그인 모달, 로그인 로직)

Choi-Jiwon-38 committed:
chore: kakao sdk type declare 추가

- dev package로 추가하는 것도 좋지만 카카오 sdk 타입 지정이 목적이어서 파일 추가하는 것으로 해결

Choi-Jiwon-38 committed:
feat: 구글 로그인 api 추가

Choi-Jiwon-38 committed:
remove: 미사용 service 제거(향후 리팩토링하여 재사용 가능성 O)

Choi-Jiwon-38 committed:
fix: `redirect_uri` 하드코딩 수정

Choi-Jiwon-38 committed:
Merge branch 'develop' into feature/1-social-accounts

Choi-Jiwon-38 commented:
https://github.com/KMU-WINK/workinkorea/pull/77 - `온보딩 페이지 개발`에서 온보딩 페이지의 경로가 `/onborading`으로 지정된 것으로 파악하여 소셜 로그인 측 redirect url을 변경해두겠습니다.

Choi-Jiwon-38 committed:
fix: 비회원 로그인 `RedirectResponse` 경로 수정

- 온보딩 페이지의 경로가 `/onboarding`으로 결정되어 이를 반영합니다.

Choi-Jiwon-38 committed:
Merge branch 'develop' into feature/1-social-accounts
